cluster_management/src/main/java/com/pinterest/rocksplicator/spectator/ConfigGeneratorClusterSpectatorImpl.java

```diff
@@ -169,8 +169,13 @@ public synchronized void start() throws Exception {
     if (this.clusterShardMapAgent == null) {
       if (shardMapZkSvr != null && !shardMapZkSvr.isEmpty()) {
         if (shardMapDownloadDir != null && !shardMapDownloadDir.isEmpty()) {
+          /**
+           * There is no need to share the zk client, as number of spectator instances is expected
+           * to be really small and total number of zk clients created will be linear to number of
+           * participant clusters being watched in total by all spectator instances...
+           */
           this.clusterShardMapAgent =
-              new ClusterShardMapAgent(shardMapZkSvr, clusterName, shardMapDownloadDir);
+              new ClusterShardMapAgent(shardMapZkSvr, null, clusterName, shardMapDownloadDir);
```

Comment:
ClusterShardMapAgent is used as sidecar on Participant to receive latest shardMap generated by Spectator, right?
If so, why does Spectator needs an ClusterShardMapAgent as well?

Reply:
Inorder to be able to perform leaf-handoff analysis. The delay from publishing to zk and then the receiving those updates on client side are measured on Spectator for the purpose of leader-handoff analysis. Although, currently the leader handoff is not integrated in distributed spectator.